sw/host/opentitanlib/src/crypto/rsa.rs

```diff
@@ -0,0 +1,255 @@
+fixed_size_bigint!(Exponent, EXPONENT_BIT_LEN);
+fixed_size_bigint!(Signature, at_most SIGNATURE_BIT_LEN);
+fixed_size_bigint!(RR, at_most RR_BIT_LEN);
+fixed_size_bigint!(N0Inv, at_most N0INV_BIT_LEN);
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Invalid public key")]
+    InvalidPublicKey,
+    #[error("Invalid DER file: {der}")]
+    InvalidDerFile {
+        der: PathBuf,
+        #[source]
+        source: anyhow::Error,
+    },
+    #[error("Read failed: {0}")]
+    ReadFailed(PathBuf),
+    #[error("Write failed: {0}")]
+    WriteFailed(PathBuf),
+    #[error("Generate failed")]
+    GenerateFailed,
+    #[error("Invalid signature")]
+    InvalidSignature,
+    #[error("Sign failed")]
+    SignFailed,
+    #[error("Verification failed")]
+    VerifyFailed,
+    #[error("Failed to compute key component")]
+    KeyComponentComputeFailed,
+}
+
+/// Ensure the components of `key` have the correct bit length.
+fn validate_key(key: impl rsa::PublicKeyParts) -> Result<()> {
+    if key.n().bits() != MODULUS_BIT_LEN || key.e().bits() != EXPONENT_BIT_LEN {
+        bail!(Error::InvalidPublicKey)
+    } else {
+        Ok(())
+    }
+}
+
+/// RSA Public Key used in OpenTitan signing operations.
+///
+/// This is a wrapper for handling RSA public keys as they're used in OpenTitan images.
+#[derive(Debug)]
+pub struct RsaPublicKey {
+    key: rsa::RsaPublicKey,
+}
+
+impl RsaPublicKey {
+    /// Construct a new public key with modulus = n and e = 65537.
+    pub fn new(n: Modulus) -> Result<RsaPublicKey> {
+        Ok(RsaPublicKey {
+            key: rsa::RsaPublicKey::new(
+                BigUint::from_bytes_le(n.to_le_bytes().as_slice()),
+                BigUint::from(65537u32),
+            )
+            .map_err(|_| Error::InvalidPublicKey)?,
+        })
+    }
+
+    /// Construct a new public key from a PKCS1 encoded DER file.
+    pub fn from_pkcs1_der_file<P: Into<PathBuf>>(der_file: P) -> Result<RsaPublicKey> {
+        let der_file = der_file.into();
+        match rsa::RsaPublicKey::read_public_key_der_file(&der_file) {
+            Ok(key) => {
+                validate_key(&key)?;
+                Ok(Self { key })
+            }
+            Err(err) => bail!(Error::InvalidDerFile {
+                der: der_file,
+                source: anyhow!(err),
+            }),
+        }
+    }
+
+    /// Write public key to a PKCS1 encoded DER file.
+    pub fn to_pkcs1_der_file<P: Into<PathBuf>>(&self, der_file: P) -> Result<()> {
+        let der_file = der_file.into();
+        self.key
+            .write_public_key_der_file(&der_file)
+            .map_err(|_| Error::WriteFailed(der_file))?;
+        Ok(())
+    }
+
+    /// Extract the public key components from a given private key.
+    pub fn from_private_key(private_key: &RsaPrivateKey) -> Self {
+        Self {
+            key: rsa::RsaPublicKey::from(&private_key.key),
+        }
+    }
+
+    /// Bit length for this key.
+    pub fn key_num_bits(&self) -> usize {
```

Comment:
```suggestion
    pub fn modulus_num_bits(&self) -> usize {
```

I assume we need this method to check the bit length before calling `modulus()`. If not, consider removing it.